utilities/fault_injection_fs.cc

```diff
@@ -606,11 +606,28 @@ IOStatus TestFSSequentialFile::Read(size_t n, const IOOptions& options,
     return s;
   }
 
+  if (read_pos_ > target_read_pos_) {
+    // We must have previously read buffered data (unsynced) not written to
+    // target. Either there has been a sync since then, in which case we can
+    // get target up to the last read_pos_, or there hasn't been a sync and
+    // we continue to serve reads from the buffer of unsynced data.
+    s = target()->Skip(read_pos_ - target_read_pos_);
+    if (s.ok()) {
+      target_read_pos_ = read_pos_;
+    } else {
+      // Wasn't synced. That's ok.
+      s = IOStatus::OK();
+    }
+  }
+
   s = target()->Read(n, options, result, scratch, dbg);
   if (!s.ok()) {
     return s;
   }
 
+  assert(result->size() == 0 || target_read_pos_ == read_pos_);
+  target_read_pos_ += result->size();
```

Comment:
What if the file is synced here? It seems we may not read all unsynced data since the buffer will be cleared before we try to read from it. Should we lock mutex_ before target()->Read() above?

Reply:
Thanks for catching this! I've almost completely revamped the approach to make it resilient to this race without killing performance or making the implementation or testing a nightmare.